pypesto/objective/finite_difference.py

```diff
@@ -1,13 +1,215 @@
+            if (
+                self.update_condition == FDDelta.SPACE and
+                np.sum((x - self.x0)**2) <= 0.5 * np.sqrt(len(x))
+            ):
+                return
+            if (
+                self.update_condition == FDDelta.STEPS and
+                (self.steps-1) % self.max_steps != 0
+            ):
+                return
+
+        # update reference point
+        self.x0 = x
+        # actually update
+        self._update(x=x, fval=fval, fun=fun, fd_method=fd_method)
+
+        if self.delta.shape != x.shape:
+            # this should not happen
+            raise ValueError("Shape mismatch.")
+
+    def _update(
+        self,
+        x: np.ndarray,
+        fval: Union[float, np.ndarray],
+        fun: Callable,
+        fd_method: str,
+    ) -> None:
+        """Actually update. Wants to be called in `update` explicitly.
+
+        Run FDs with various deltas and pick the ones, separately for each
+        parameter, with the best stability properties.
+
+        The parameters are the same as for
+        :func:`pypesto.objective.finite_difference.FDDelta.update`.
+        """
+
+        # calculate gradients for all deltas for all parameters
+        nablas = []
+        # iterate over deltas
+        for delta in self.test_deltas:
+            # calculate Jacobian with step size delta
+            delta_vec = delta * np.ones_like(x)
+            nabla = fd_nabla_1(
+                x=x, fval=fval, f_fval=fun, delta_vec=delta_vec,
+                fd_method=fd_method)
+
+            nablas.append(nabla)
+
+        # shape (n_delta, n_par, ...)
+        nablas = np.array(nablas)
+
+        # The stability vector is the the absolute difference of Jacobian
+        #  entries towards smaller and larger deltas, thus indicating the
+        #  change in the approximation when changing delta.
+        # This is done separately for each parameter. Then, for each the delta
+        #  with the minimal entry and thus the most stable behavior
+        #  is selected.
+        stab_vec = np.full(shape=nablas.shape, fill_value=np.nan)
+        stab_vec[1:-1] = 0.5 * (
+            np.abs(nablas[2:] - nablas[1:-1]) +
+            np.abs(nablas[1:-1] - nablas[:-2])
+        )
+        # on the edge, just take the single neighbor
+        stab_vec[0] = np.abs(nablas[1] - nablas[0])
+        stab_vec[-1] = np.abs(nablas[-1] - nablas[-2])
+
+        # if the function is tensor-valued, consider the mean over all entries
+        # there may be more intelligent ways
+        if stab_vec.ndim > 2:
+            # flatten all dimensions > 1
+            stab_vec = stab_vec.reshape(
+                stab_vec.shape[0], stab_vec.shape[1], -1).mean(axis=2)
+
+        # minimum delta index for each parameter
+        min_ixs = np.argmin(stab_vec, axis=0)
+
+        # extract optimal deltas per parameter
+        delta_opt = np.array([self.test_deltas[ix] for ix in min_ixs])
+        self.delta = delta_opt
```

Comment:
Personally, I'm a fan of having things explicit and not craming everything into one line or LC. However, one could think of writing `self.delta = np.array([self.test_deltas[ix] for ix in min_ixs])` ... ;)